### ISSUE TITLE
Problems running tests using dap-java

### ISSUE THREAD

**Issue by SerialVelocity** [yyoncho/dap-mode#10]
Hey,

Great work on dap-mode so far! I thought I'd leave a message with all the issues I encountered though:
- When running `dap-java-run-test-class`, it fails with:
  ```
  compilation-start: Wrong type argument: stringp, :program-to-start
  ```
  Fixed by replacing `(first ...)` with `(plist-get ... :program-to-start)` in `dap-java- run-test-class` (to make it mirror `dap-java-run-test-method`).

- When running `dap-java-run-test-method`, it fails with:
  ```
  org.junit.platform.commons.util.PreconditionViolationException: No arguments were supplied to the ConsoleLauncher
  ```
  This is because the class path isn't being passed to the `compile` command.
  Fixed with a simple `(mapc (-lambda ((env . value)) (setenv env value)) (plist-get params :environment-variables)` stolen from your debug codepath.
- The working directory for running tests is wrong. It looks like the working directory used for running the test is where the `.java` file is located. It should be the project root (not necessarily the workspace root).
- The junit runner outputs colours. It would be nice if you used a buffer which supports them. Maybe this is just me needing to use something like `ansi-color`? (would be nice to have support built-in though)

Best,
Ben

P.S. I *love* that the output is live. I really hated that both emacs and eclipse hung when using eclim. ❤️

**Comment by yyoncho** [comment]
Your feedback is greatly appreciated! Feel free to comment and provide feedback for anything that does not look prod quality.

As for the bug report, I will check what is wrong - I am using the debug version of the methods(which do have test coverage). Probably these were broken by the recent changes.

**Comment by SerialVelocity** [comment]
Thanks!

Another thing I've found is the import order from my `org.eclipse.jdt.ui.prefs` isn't used. I can live with that even though it is a bit annoying.

The bit that is actually a bug is that the server throws an NPE when trying to organise my imports when I have `(setq lsp-java-import-order '())`

The new configuration log line shows:
```
..., importOrder=null, ...
```

It probably has to be completely excluded from the configuration?

**Comment by yyoncho** [comment]
Hm, can you set your import order in `lsp-java-import-order` or you have custom import order per project? 

> It probably has to be completely excluded from the configuration?

Or `lsp-java` has to convert nil to an empty array.

**Comment by yyoncho** [comment]
I have pushed - https://github.com/yyoncho/dap-mode/commit/c517083ae8bb927cf33ee26d00d5c00eb028c8df
See comment inline:

> When running dap-java-run-test-class, it fails with:
> 
> compilation-start: Wrong type argument: stringp, :program-to-start
> Fixed by replacing (first ...) with (plist-get ... :program-to-start) in dap-java- run-test-class (to make it mirror dap-java-run-test-method).
> 
> When running dap-java-run-test-method, it fails with:
> 
> org.junit.platform.commons.util.PreconditionViolationException: No arguments were supplied to the ConsoleLauncher
> This is because the class path isn't being passed to the compile command.
> Fixed with a simple (mapc (-lambda ((env . value)) (setenv env value)) (plist-get params :environment-variables) stolen from your debug codepath.
> 

All these are fixed. I have changed the code to go through the same path as the debug methods.

> The working directory for running tests is wrong. It looks like the working directory used for running the test is where the .java file is located. It should be the project root (not necessarily the workspace root).
> 

At this point the proper fix is not easy to implement, it depends on code that  I already have but it its submission is blocked by https://github.com/emacs-lsp/lsp-mode/pull/409 . I have provided a workaround - you may create your custom run command that calculates the proper cwd and sets the it in launch-args . You could do that by locating the build file(e. g. pom.xml). Later when https://github.com/emacs-lsp/lsp-mode/pull/409 is in I will change the code to calculate the root based on the current project root.

> The junit runner outputs colours. It would be nice if you used a buffer which supports them. Maybe this is just me needing to use something like ansi-color? (would be nice to have support built-in though)

It works on my machine - I believe it should work out of the box for compile buffers, I guess the problem is related to your configuration. Later I will check whether it is related to my spacemacs config.

**Comment by yyoncho** [comment]
> Maybe this is just me needing to use something like ansi-color? (would be nice to have support built-in though)

I guess you might need this(source: https://stackoverflow.com/questions/13397737/ansi-coloring-in-compilation-mode)
``` elisp
(require 'ansi-color)
(defun colorize-compilation-buffer ()
  (toggle-read-only)
  (ansi-color-apply-on-region compilation-filter-start (point))
  (toggle-read-only))
(add-hook 'compilation-filter-hook 'colorize-compilation-buffer)
```
I am a bit reluctant to include that in the `dap-mode` since this setting is shared for all compilation buffers and it looks like doing this is not in the spirit of emacs.

**Comment by SerialVelocity** [comment]
> All these are fixed. I have changed the code to go through the same path as the debug methods.

Thanks for the quick turnaround. I think there's a bug though. I'm getting:
```
Debugger entered--Lisp error: (void-variable cwd)
  (let* ((:skip-debug-session (plist-get launch-args :name)) (:program-to-start (plist-get launch-args :cwd)) (:type (plist-get launch-args :wait-for-port)) (:port (plist-get launch-args :request)) (:hostName (plist-get launch-args :environment-variables)) (default-directory cwd)) (mapc (function (lambda (input0) (let* ((--dash-source-31-- input0) (env (car-safe ...)) (value --dash-source-31--)) (setenv env value)))) environment-variables) (if program-to-start (progn (compile program-to-start))) (if wait-for-port (progn (dap--wait-for-port host port))) (if skip-debug-session nil (let ((debug-session (dap--create-session launch-args)) (workspace lsp--cur-workspace) (breakpoints (dap--get-breakpoints lsp--cur-workspace))) (dap--send-message (dap--initialize-message type) (dap--session-init-resp-handler debug-session (function (lambda (initialize-result) (let* ... ... ...) (dap--send-message ... ... debug-session)))) debug-session) (dap--set-cur-session debug-session) (setq dap--debug-configuration (cons (cons name launch-args) dap--debug-configuration)) (run-hook-with-args (quote dap-session-created-hook) debug-session))))
  dap-start-debugging((:program-to-start "java -jar ~/.emacs.d/eclipse.jdt.ls/runner/junit-platform-console-standalone.jar -cp $JUNIT_CLASS_PATH -m com.class.name#testName" :environment-variables (("JUNIT_CLASS_PATH" . "stupidly long classpath")) :cwd "<path>" :skip-debug-session t))
  dap-java-run-test-method()
  funcall-interactively(dap-java-run-test-method)
  call-interactively(dap-java-run-test-method nil nil)
  command-execute(dap-java-run-test-method)
```

> I have provided a workaround

Thanks! I'll try to use it shortly.

> I guess you might need this(source: stackoverflow.com/questions/13397737/ansi-coloring-in-compilation-mode)

Thanks, I got that working just after I posted my message. It would be nice to have a link from the README though for other people encountering the same problem.

> I am a bit reluctant to include that in the dap-mode since this setting is shared for all compilation buffers and it looks like doing this is not in the spirit of emacs.

It's a shame the buffer isn't named otherwise it could be done just for dap buffers

**Comment by SerialVelocity** [comment]
> Hm, can you set your import order in lsp-java-import-order or you have custom import order per project?

It is custom per project but most of the projects I work on at the moment don't have an ordering. It's just alphabetical so no grouping (hence the `'()`).

I suggested omitting it because I don't know whether the server is overriding what is in the `org.eclipse.jdt.ui.prefs` file with what you are passing though.

**Comment by yyoncho** [comment]
You have to upgrade dash. You do not have the proper version, it's too old. I will ping you with more details on imports tomorrow my time.

**Comment by SerialVelocity** [comment]
Thanks, that did indeed fix it. I did have a quick look at the changes to `-let*` but couldn't see any important fixes so I ignored that possibility. I was wrong.

To work around the import issue for now, I set it to `'("non-existent")`

**Comment by yyoncho** [comment]
Ok, I am closing this for now:
* The organize import will be addressed in `lsp-java`
* I will improve project root calculation logic once `lsp-mode` fix is in.

**Comment by yyoncho** [comment]
@SerialVelocity pushed https://github.com/emacs-lsp/lsp-java/commit/4e4a67961777555bc37c8ddf94833520398935ae . Now you can pass nil for the lists and it will be sent as empty array over the wire.
